### PR TITLE
Problem: packet get relayed even estimated gas is higher than max gas

### PR DESCRIPTION
**Pull request by mmsqe**
to align estimate fee behaviour with [hermes](https://github.com/informalsystems/hermes/blob/77423055c25b968fb3b0f704a76a955458262ee9/crates/relayer/src/chain/cosmos/estimate.rs#L76)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -22,6 +22,7 @@
 * [\#1221](https://github.com/cosmos/relayer/pull/1221) Update cometbft to v0.37.2 and ibc-go to v7.2.0.
 * [\#1226](https://github.com/cosmos/relayer/pull/1226) Avoid invalid Bech32 prefix error in parallel tests when sdk Config get overwritten by each other in single process.
 * [\#1231](https://github.com/cosmos/relayer/pull/1231) Reduce get bech32 prefix when get signer.
+* [\#](https://github.com/cosmos/relayer/pull/) Avoid packet get relayed when estimated gas is higher than max gas.
```
Comment (mmsqe) [inline CHANGELOG.md]:
```suggestion
* [\#1302](https://github.com/cosmos/relayer/pull/1302) Avoid packet get relayed when estimated gas is higher than max gas.
```

File: relayer/chains/cosmos/tx.go [hunk 2]
```diff
@@ -267,8 +266,6 @@ func (cc *CosmosProvider) SendMsgsWith(ctx context.Context, msgs []sdk.Msg, memo
 		if err != nil {
 			return nil, err
 		}
-
-		adjusted = uint64(float64(adjusted) * cc.PCfg.GasAdjustment)
```
Comment (jtieri) [inline relayer/chains/cosmos/tx.go:271]: good catch!

File: relayer/chains/cosmos/tx.go [hunk 3]
```diff
@@ -1676,21 +1673,16 @@ func (cc *CosmosProvider) PrepareFactory(txf tx.Factory, signingKey string) (tx.
...
-	if math.IsInf(gas, 1) {
-		return 0, fmt.Errorf("infinite gas used")
-	}
-	// Bound the gas estimate by the max_gas option
-	if cc.PCfg.MaxGasAmount > 0 {
-		gas = math.Min(gas, float64(cc.PCfg.MaxGasAmount))
+	if cc.PCfg.MaxGasAmount > 0 && gasUsed > cc.PCfg.MaxGasAmount {
+		return 0, fmt.Errorf("estimated gas %d is higher than max gas %d", gasUsed, cc.PCfg.MaxGasAmount)
 	}
+	gas := cc.PCfg.GasAdjustment * float64(gasUsed)
```
Comment (jtieri) [inline relayer/chains/cosmos/tx.go]: this change looks fine but do you think we should leave the check from `math.IsInf` in place? It is possible that a user accidentally, or without realizing the implications, sets the value of `GasAdjustment` to some insanely large integer value and then we could end up with `MaxFloat64` from `cc.PCfg.GasAdjustment * float64(gasUsed)`

**Review by jtieri — COMMENTED** [review]
i think we should possibly add the check back that i point out in the review but other than that this lgtm

**Review by jtieri — APPROVED** [review]
thanks!

**Pull request merged**